crates/kitsune_p2p/types/src/tx2/tx2_promote.rs

```diff
@@ -0,0 +1,701 @@
+                Err(e) => {
+                    // unable to resolve incoming channel
+                    // shut down the connection
+
+                    let reason = format!("{:?}", e);
+
+                    // TODO - standardize codes?
+                    close_inner(inner, 500, &reason).await;
+
+                    // exit the loop
+                    return;
+                }
+                Ok(c) => c,
+            };
+            loop {
+                let r = chan
+                    .read(KitsuneTimeout::from_millis(MAX_READ_TIMEOUT))
+                    .await;
+                let (msg_id, data) = match r {
+                    Err(e) if *e.kind() == KitsuneErrorKind::Closed => {
+                        // this channel was closed - exit the loop
+                        // allowing another channel to be processed.
+                        return;
+                    }
+                    Err(e) => {
+                        // unrecoverable error - shut down the connection
+
+                        let reason = format!("{:?}", e);
+
+                        // TODO - standardize codes?
+                        close_inner(inner, 500, &reason).await;
+
+                        // exit the loop
+                        return;
+                    }
+                    Ok(r) => r,
+                };
+
+                let data = EpIncomingData {
+                    url: url.clone(),
+                    con: con.clone(),
+                    msg_id,
+                    data,
+                };
+                if logic_hnd.emit(EpEvent::IncomingData(data)).await.is_err() {
+                    break;
+                }
+            }
+        })
+        .boxed();
+
+    let write_fut = async move {
+        let limit = Arc::new(Semaphore::new(CHANNELS_PER_CONNECTION));
+        loop {
+            let permit = match limit.clone().acquire_owned().await {
+                Err(_) => {
+                    // we only get errors here when our endpoint has closed
+                    // we can safely just exit this loop
+                    return;
+                }
+                Ok(p) => p,
+            };
+            let writer = match raw_con
+                .out_chan(KitsuneTimeout::from_millis(MAX_READ_TIMEOUT))
+                .await
+            {
+                Err(e) => {
+                    // we were not able to create an outgoing channel
+                    // clean up the connection.
+
+                    let reason = format!("{:?}", e);
+
+                    // TODO - standardize codes?
+                    close_inner(inner, 500, &reason).await;
+
+                    // exit the loop
+                    return;
+                }
+                Ok(c) => c,
+            };
+            writer_bucket.release(WriteChan {
+                _permit: permit,
+                writer,
+            });
+        }
+    };
+
+    let _ = futures::future::join(recv_fut, write_fut).await;
```

Comment:
Might be a non-issue but there's a few silent errors in here that might make debugging hard.

Reply:
I'm fairly confident I've promoted all the real errors to the connection close event, everything else only happens on endpoint shutdown which are appropriate to ignore.